### DISCUSSION TITLE
XMLSec 1.3.3

### DISCUSSION THREAD

**Discussion by lsh123** [lsh123/xmlsec#750] [Announcements]
The [XML Security Library 1.3.3](https://www.aleksey.com/xmlsec-rc/download.html) release includes the following changes:

- (xmlsec-core) Disabled KeyValue and DEREncodedKeyValue XML nodes by default. Use the '--enabled-key-data' option for the xmlsec command line utility or update the 'keyInfoCtx->enabledKeyData' parameter if you need to re-enable these nodes (also see [question 3.5 in the FAQ](https://www.aleksey.com/xmlsec-rc/faq.html#section_3_5)).
- (xmlsec-core) Removed '--enable-size-t' ('size_t' for MSVC builds) option and made 'xmlSecSize' to always be the same as 'size_t'.
- (xmlsec-core) Removed previously deprecated functions, defines, etc.
- (xmlsec-core) Fixed build for libxml2 v2.12.0.
- (xmlsec-openssl) Removed support for OpenSSL 1.1.0 ([end of life in Aug 2016](https://endoflife.date/openssl)). The minimum OpenSSL supported version is 1.1.1; the version 3.0.0 or greater is recommended.
- (xmlsec-nss) Added runtime check for the enabled algorithms in NSS.
- (xmlsec-mscrypto) Removed NT4 support.
- Several other small fixes (see [more details](https://github.com/lsh123/xmlsec/commits/master)).

Thanks for bug reports and PRs!

Aleksey

<hr /><em>This discussion was created from the release <a href='https://github.com/lsh123/xmlsec/releases/tag/1.3.3'>XMLSec 1.3.3</a>.</em>

**Comment by vmiklos** [comment]
Hm, I hit a problem with commit 8fc21b2af04161138c298105b0fadea06cbd9185 while updating the bundled xmlsec on the libreoffice side, where we build against a debug nss in debug builds, i.e. asserts actually can fail (which is not something I tested with odfsig).

Here is the backtrace of the failure I see:

```
            #1  0x00007ffff77503e5 in abort () at /lib64/libc.so.6
            #2  0x00007ffff679ca90 in PR_Assert (s=0x7ffff44fbfe5 "oidmechhash != NULL", file=0x7ffff44fbfb6 "secoid.c", ln=2140) at ../../../../pr/src/io/prlog.c:571
            #3  0x00007ffff44eaae0 in SECOID_FindOIDByMechanism (mechanism=307) at secoid.c:2140
            #4  0x00007ffff6660c65 in PK11_MechanismToAlgtag (type=307) at pk11mech.c:1745
            #5  0x00007fffe874e4ea in xmlSecNssCryptoCheckMechanism (type=307) at crypto.c:68
            #6  0x00007fffe874ec44 in xmlSecNssUpdateAvailableCryptoTransforms (functions=0x7fffe87ffbc0 <functions>) at crypto.c:401
            #7  0x00007fffe874f13d in xmlSecNssInit () at crypto.c:552
            #8  0x00007fffe86b2c36 in initXmlSec() () at /home/vmiklos/git/libreoffice/core/xmlsecurity/source/xmlsec/xmlsec_init.cxx:42
```

Does this ring a bell for you? This is against nss-3.90, perhaps the code was only tested against some older nss or only against nss with asserts disabled?

If it's useful, I can try to collect a bit more info, like what is the value of type when the called `PK11_MechanismToAlgtag()` triggers an assertion failure or so. Thanks.

Edit: ah, the type is there already, 307, which is 0x133, which seems to be CKM_DES3_CBC. 

Edit 2: the below fixes the assertion failure for me:

```
diff --git a/src/nss/crypto.c b/src/nss/crypto.c
index 6455ec72..c23ff66e 100644
--- a/src/nss/crypto.c
+++ b/src/nss/crypto.c
@@ -361,8 +361,16 @@ xmlSecCryptoGetFunctions_nss(void) {
 
 static void
 xmlSecNssUpdateAvailableCryptoTransforms(xmlSecCryptoDLFunctionsPtr functions) {
+    SECStatus rv;
+
     xmlSecAssert(functions != NULL);
 
+    rv = SECOID_Init();
+    if (rv != SECSuccess) {
+        xmlSecNssError("SECOID_Init", NULL);
+        return;
+    }
+
     /******************************* AES ********************************/
     /* cbc */
     if (xmlSecNssCryptoCheckAlgorithm(SEC_OID_AES_128_CBC) == 0) {
```

Looks like debug nss asserts that the OID hashtable is already initialized, but if asserts are off, then it will just silently initialize the hashtable, so nothing bad happens. Should I send this in as a proper PR or you have a better idea how to avoid the problem? Thanks.

Edit 3: probably the intention of the nss assert is that auto-init is bad, beause the init and shutdown should be called as a pair, i.e. SECOID_Init() before this new code and SECOID_Shutdown() at the end, so the above patch is just a PoC, would still leak memory.

**Comment by lsh123** [comment]
Sorry for delay with reply -- I am traveling for the next few days. Looks like you already found the reason -- NSS was not initialized before xmlsec-nss init was called. I can add this change from above but the proper solution is to init NSS before xmlsec.

**Reply by lsh123** [reply]
https://github.com/lsh123/xmlsec/pull/754

**Reply by vmiklos** [reply]
I see, thanks for the fast fix! It solves the problem I see, later I'll see to also do a an NSS init/deinit around xmlsec on our side to have this fixed on both sides.
